pkg/controllers/scheduling/schedule.go

```diff
@@ -0,0 +1,160 @@
+package scheduling
+
+import (
+	"context"
+	"fmt"
+	"reflect"
+	"sort"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/labels"
+	"k8s.io/apimachinery/pkg/selection"
+
+	clusterclient "github.com/open-cluster-management/api/client/cluster/clientset/versioned"
+	clusterlisterv1alpha1 "github.com/open-cluster-management/api/client/cluster/listers/cluster/v1alpha1"
+	clusterapiv1 "github.com/open-cluster-management/api/cluster/v1"
+	clusterapiv1alpha1 "github.com/open-cluster-management/api/cluster/v1alpha1"
+)
+
+type scheduleFunc func(
+	ctx context.Context,
+	placement *clusterapiv1alpha1.Placement,
+	clusters []*clusterapiv1.ManagedCluster,
+	clusterClient clusterclient.Interface,
+	placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister,
+) (*scheduleResult, error)
+
+type scheduleResult struct {
+	scheduled   int
+	unscheduled int
+}
+
+func schedule(
+	ctx context.Context,
+	placement *clusterapiv1alpha1.Placement,
+	clusters []*clusterapiv1.ManagedCluster,
+	clusterClient clusterclient.Interface,
+	placementDecisionLister clusterlisterv1alpha1.PlacementDecisionLister,
+) (*scheduleResult, error) {
+	// filter clusters with cluster predicates
+	feasibleClusters, err := matchWithClusterPredicates(placement.Spec.Predicates, clusters)
+	if err != nil {
+		return nil, err
+	}
+
+	// select clusters and generate cluster decisions
+	decisions := selectClusters(placement, feasibleClusters)
+	scheduled := 0
+	for _, decision := range decisions {
+		if len(decision.ClusterName) > 0 {
+			scheduled++
+		}
+	}
+
+	// bind the cluster decisions into placementdecisions
+	err = bind(ctx, placement, decisions, clusterClient, placementDecisionLister)
+	if err != nil {
+		return nil, err
+	}
+
+	return &scheduleResult{
+		scheduled:   scheduled,
+		unscheduled: len(decisions) - scheduled,
+	}, nil
+}
+
+// makeClusterDecisions selects clusters based on given cluster slice and then creates
+// cluster decisions. Empty cluster decisions will be appended if the number of selected
```

Comment:
Why does the API add empty cluster decisions?

Reply:
The logic to create empty cluster decisions has been removed. Empty cluster decisions are no longer needed because the scheduling controller does not schedule based on unscheduled decisions (the empty decisions) in the status of placementdecisions any more.